Refactor conngen module to be a ConnBuilder

jougs opened the pull request:
This PR refactors the Connection Generator interface for NEST to be a regular `ConnBuilder` called `conngen` instead of being an extension module called `conngen`, which used to provide the `CGConnect` function. This addresses one of the action items of  #1717.

jougs commented on nestkernel/conn_builder_conngeninterface.cpp:
```diff
@@ -20,158 +20,133 @@
-        dummy_params,
-        params[ d_idx ],
-        params[ w_idx ] );
+      delay_ = new ScalarDoubleParameter( params[ d_idx ], num_threads );
+      weight_ = new ScalarDoubleParameter( params[ w_idx ], num_threads );
```
@hakonsbm: I was hoping that setting `weight_` and `delay_` here would have the effect that the subsequent call to `single_connect` will just use these values for establishing the connections. Unfortunately, this seems to not be the case and the connections are just created with defaults.

A completely orthogonal idea would of course be to directly call `ConnectionManager::connect()` here and not use any `Parameters` and `single_connect()` at all. For that I'd probably would need to get some values (e.g. the synapse model) from the *syn_spec*.

So, the point is that, the whole mechanism for `Parameters` and how to implement a `ConnBuilder` has apparently become a bit too complex for me and I'm a bit at loss here. Could you maybe give me some advice? I suggest to have a dedicated session on this during the hackathon next week in which we could also discuss a bit about how to document all this.

jougs commented on nestkernel/conn_builder_conngeninterface.cpp:
```diff
@@ -20,158 +20,133 @@
+  const int num_parameters = cg_->arity();
   if ( num_parameters == 0 )
   {
+    weight_ = delay_ = 0;
+
```
Remove to use defaults

jougs commented on doc/installation/compilation_options.rst:
```diff
@@ -137,6 +133,21 @@ already contains an example implementation that should work with the
+-----------------------
+
+In order to allow NEST to create connections using external libraries,
+it provides an interface for the Connection Generator Interface from
```
```suggestion
it provides support for the Connection Generator Interface from
```

jougs commented on nestkernel/conn_builder.cpp:
```diff
@@ -58,9 +58,10 @@ nest::ConnBuilder::ConnBuilder( NodeCollectionPTR sources,
   , use_pre_synaptic_element_( false )
   , use_post_synaptic_element_( false )
+  , parameters_requiring_skipping_()
+  , synapse_model_id_( kernel().model_manager.get_synapsedict()->lookup( "static_synapse" ) )
```
@stinebuu: Not sure why I have this one. Maybe it should go away?

jougs replied:
Removed in e40c193

jougs commented on pynest/examples/csa_spatial_example.py:
```diff
@@ -28,25 +28,26 @@
+.. [1] Djurfeldt M, Davison AP and Eppler JM (2014). Efficient generation of
+       connectivity in neuronal networks from simulator-independent
+       descriptions. Front. Neuroinform.
        http://dx.doi.org/10.3389/fninf.2014.00043
```
```suggestion
       https://doi.org/10.3389/fninf.2014.00043
```

stinebuu commented on nestkernel/conn_builder_conngen.cpp:
```diff
@@ -20,158 +20,140 @@
   }
-  const index synmodel_id = static_cast< index >( synmodel );
-  DictionaryDatum dummy_params = new Dictionary();
+}
```
You should probably also check if `syn_specs.size() > 1` and throw an error if this is the case. Collocation of synapses should not be possible if we use the `conngen` rule.

jougs replied:
Ah, yes. Now that you say it, I remember that we talked about this! Fixed in 824caee56. Could you maybe provide me with a code snipped that creates collocated synapses, so I can write a test that checks if the error reporting works? Thanks!

stinebuu replied:
A simple example would be:
```
node = nest.Create('iaf_psc_alpha', 3)
nest.Connect(node, node, syn_spec=nest.CollocatedSynapses({'weight': -2.}, {'weight': 2.}))
```

stinebuu replied:
If you want to test without weight, you could use other parameters than weights, like `synapse_model` and other parameters that can be set for the individual synapse types.
```
node = nest.Create('iaf_psc_alpha', 3)
nest.Connect(node, node, syn_spec=nest.CollocatedSynapses({'synapse_model': 'stdp_synapse', 'alpha': 3.}, {'synapse_model': 'static_synapse'}))
```

jougs replied:
I've added a test in 500a7412b.

stinebuu commented on nestkernel/conn_builder_conngen.cpp:
```diff
@@ -20,158 +20,140 @@
+    if ( syn_specs[ 0 ]->known( names::weight ) or syn_specs[ 0 ]->known( names::delay ) )
+    {
+      throw BadProperty( "Properties weight and delay cannot be specified in syn_spec if the "
+			 "ConnectionGenerator has values." );
```
Is the indentation right here? It looks a bit weird, but it might just be me :)

jougs replied:
Fixed in c2b4eade0.

stinebuu commented on nestkernel/conn_builder_conngen.cpp:
```diff
@@ -20,158 +20,140 @@
+    if ( not( d_idx_is_0_or_1 and w_idx_is_0_or_1 and indices_differ ) )
     {
-      throw BadProperty( "w_idx and d_idx have to differ and be either 0 or 1." );
+      throw BadProperty( "d_idx and w_idx have to be either 0 or 1 and cannot be the same." );
```
It might be better to use `delay index and weight index` instead of `d_idx and w_idx` in the message.

jougs replied:
What about
```suggestion
      throw BadProperty( "The indices for weight and delay have to be either 0 or 1 and cannot be the same." );
```

stinebuu replied:
Looks good!

stinebuu reviewed (changes requested):
Nice! I have some comments and suggestions, but they shouldn't be too big.

jougs commented on examples/nest/conngen_interface.sli:
```diff
@@ -22,19 +22,17 @@
+Name: testsuite::test_conngen_interface - Test runability of Connect variant with conngen rule
 
-   Synopsis: (test_conngen_interface) run
+Synopsis: (test_conngen_interface) run
```
Once upon a time, in a land far far away, we all believed that a C++ implementation of the CSA is on its way and will become available *very soon*. It was at that time that I created this file, which was meant to be a *test*. Then reality came (in total contrast to that C++ implementation) and I moved it to `examples` to keep it (hope dies last and all that...).

So much for history (and fairy tales). I have pushed a slightly updated version of the file in f844959be.

stinebuu replied:
Makes sense, thanks for updating :)

jougs commented on pynest/nest/tests/test_csa.py:
```diff
@@ -167,44 +113,61 @@ def test_CSA_OneToOne_synmodel(self):
             self.assertEqual(len(conns), 1)
             self.assertEqual(conns[0]["target"], targets[i].get('global_id'))
```
Much nicer, thanks! Fixed in ccfae0d05.

sarakonradi reviewed (changes requested):
Thanks! I've added some documentation-related suggestions inline.

jougs commented on doc/guides/connection_management.rst:
```diff
@@ -194,10 +194,51 @@ Example:
+
+To allow the generation of connectivity by means of an external
+library, NEST supports the Connection Generator Interface (`Djurfeldt
+et al., 2014 <https://doi.org/10.3389/fninf.2014.00043>`_).
```
Good idea. Done in d5802c9a5.

jougs commented:
@stinebuu, @sarakonradi: I've addressed all your points in my recent commits. Not sure if we really need the review by @hakonsbm. If you think we're good without, feel free to say so or remove the request of review.

stinebuu reviewed (approved):
Thank you!

sarakonradi reviewed (approved):
LGTM!
